pysmt/optimization/optimizer.py

```diff
@@ -138,6 +140,93 @@ def _get_lt(self, objective_formula):
             raise PysmtValueError("Invalid optimization function type: %s" % otype)
 
 
+class OptSearchInterval():
+
+    def __init__(self, goal):
+        self._obj = goal
+        self._lower = None #-INF where i found this costant?
+        self._upper = None #+INF
+        self._pivot = None
+
+    def _comparation_functions(self, objective_formula):
+        """Internal utility function to get the proper cast, LT and LE
+        function for the given objective formula
+        """
+        otype = self.environment.stc.get_type(objective_formula)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type():
+            return mgr.Int, mgr.LT, mgr.LE
+        elif otype.is_real_type():
+            return mgr.Real, mgr.LT, mgr.LE
+        elif otype.is_bv_type():
+            return (lambda x: mgr.BV(x, otype.width)), mgr.BVULT, mgr.BVULE
+        else:
+            raise PysmtValueError("Invalid optimization function type: %s" % otype)
+
+    def _make_better_than(self, goal, x):
+        cast, lt, _ = self._comparation_functions(goal.term())
+        if goal.is_minimization_goal():
+            t1 = goal.term()
+            t2 = x
+        else:
+            t1 = x
+            t2 = goal.term()
+        return lt(t1, t2)
+
+    def linear_search_cut(self):
+        """must be called always"""
+        if self._obj.is_minimization_goal():
+            bound = self._upper
+        else:
+            bound = self._lower
+
+        return self.make_better_than(self._obj, bound)
+
+    def _compute_pivot(self):
+        if self._lower is None and self._upper is None:
+            return None
+        l,u = self._lower, self._upper
+        if self._lower is None and self._upper is not None:
+            l = self._upper - (abs(self._upper) + 1)
+        elif self._lower is not None and self._upper is None:
+            u = self._lower + abs(self._lower) + 1
+        return (l + u + 1) // 2
```

Comment:
1. Is `(l + u + 1) // 2` type-correct in SMT-LIB, given that `l`,`u` can be of Int/Real/BV type?

2. If `obj.term` is of Int type, one should probably round the result of the division.

3. The `pivot` value should be validated before returning it. It should be discarded
    when it is _smaller or equal `self._lower` (minimization)_
    or _larger or equal `self._upper` (maximization)_.

4. `binary_search_cut()` expects a valid `pivot` all the time, but the current implementation
     of this method does not guarantee that. A safe pivoting value for minimization is `self._upper`
     and for maximization is `self._lower`. _(i.e. it becomes a linear step)_

Reply:
@mikand 

1. What is the concrete model value of a real variable? An infinite precision fraction or a float?

2. agreed

3./4. The previous implementation returned `None`, which was not valid. Now for Int/BV it appears to be correct. For Reals, after removing `+1`, it might depend on the type of `l,u`. I wouldn't trust the result of `(l + u)/2` to return some value in the middle for every pair of `float` values. I would trust it if `l,u` are represented as infinite precision fractions.

Reply:
@PatrickTrentin88 

in pysmt we use infinite-precision fractions (mpq_t from GMP if pygmp is installed otherwise `fractions.Fraction` from the standard library) for rationals and plain integers (that in python are arbitrary-precision) for INTs (and also as internal representation of BVs).

@arrighi79 Be careful in using these types for lower and upper bounds. You can also use assertions to check them.

Reply:
> @PatrickTrentin88
> 
> in pysmt we use infinite-precision fractions (mpq_t from GMP if gmpy2 is installed otherwise `fractions.Fraction` from the standard library) for rationals 

Out of curiosity, why do you use `mpq_t` if available and `fractions.Fraction` otherwise? It looks like they are interchangeable, since you can swap one for the other, so why bother?

Reply:
Math operations with gmpy are WAY faster than with fractions.Fraction. This is negligible if you use the python representations just for representation, but if you use them in algorithms and you do computation the difference in terms of performance is huge.